Convert PQ 4 bit code to dynamic dispatch

mdouze opened the pull request:
Summary:
Migration of the 4-bit codecs to dynamic dispatch.

The migration consists in:
- templatizing the SIMD ResultHandlers to the SIMDLevel
- instantiating the AVX2 and AVX512 code in their own files (compile units)
- removing any SIMD dependency from IndexFastScan and IndexIVFFastScan
- adding dispatching code for the SIMD code

Differential Revision: D73581633

alexanderguzhva commented:
@mdouze is it good to review? 'Facebook Internal' tests has been running for 5 days O_o

mdouze commented:
> @mdouze is it good to review? 'Facebook Internal' tests has been running for 5 days O_o

Not it's not yet ready, we have a significant TODO list to add. 
However, if you have comments, feel free to leave them.

mdouze committed:
a few more functions in distances_simd.cpp

Summary:
Model to apply dynamic dispatch to functions that are auto-vectorized.

Note that the implementation in knowhere prefers using function pointers than to do a switch. The advantage is that it is probably faster. The disadvantage is that it changes the function prototypes.

Differential Revision: D72937711

mdouze committed:
Common ancestor to top-k result handlers

Differential Revision: D73581632

mdouze committed:
Use simdlib abstraction in ScalarQuantizer implementation, split off training code, split quantizer code into headers, Make headers more independent

Summary:
Move the interface of SIMD functions to use the simdXfloat32 API to mutualize code.
Begin splitting the ScalarQuantizer.cpp
Continue splitting. Purely in header files for now.

Differential Revision: D72945865

mdouze committed:
fvec_madd simd impl

Differential Revision: D72937708

mdouze committed:
Split ScalarQuantizer code into independent parts (#4296)

Summary:
Pull Request resolved: https://github.com/facebookresearch/faiss/pull/4296

Splits the ScalarQuantizer code into parts so that the AVX2 and AVX512 can be compiled independently.

Differential Revision: D73037185

mdouze committed:
moved IndexIVFPQ and IndexPQ to dynamic dispatch (#4291)

Summary:
Pull Request resolved: https://github.com/facebookresearch/faiss/pull/4291

moved IndexIVFPQ and IndexPQ to dynamic dispatch.
Since the code was already quite modular (thanks Alex!), this boils down to make independent cpp files for the different SIMD versions.

Differential Revision: D72937709

mdouze committed:
SIMDConfig object

Differential Revision: D72937710

mdouze committed:
Convert PQ 4 bit code to dynamic dispatch (#4309)

Summary:
Pull Request resolved: https://github.com/facebookresearch/faiss/pull/4309

Migration of the 4-bit codecs to dynamic dispatch.

The migration consists in:
- templatizing the SIMD ResultHandlers to the SIMDLevel
- instantiating the AVX2 and AVX512 code in their own files (compile units)
- removing any SIMD dependency from IndexFastScan and IndexIVFFastScan
- adding dispatching code for the SIMD code

Differential Revision: D73581633

mnorris11 commented:
Merged in different PRs
